File: app/src/main/java/xyz/zedler/patrick/tack/activity/SplashActivity.java
package xyz.zedler.patrick.tack.activity;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.graphics.drawable.LayerDrawable;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.content.res.ResourcesCompat;
import com.google.android.material.color.DynamicColors;
import xyz.zedler.patrick.tack.Constants.DEF;
import xyz.zedler.patrick.tack.Constants.EXTRA;
import xyz.zedler.patrick.tack.Constants.PREF;
import xyz.zedler.patrick.tack.Constants.THEME;
import xyz.zedler.patrick.tack.R;
import xyz.zedler.patrick.tack.behavior.SystemBarBehavior;
import xyz.zedler.patrick.tack.util.PrefsUtil;
import xyz.zedler.patrick.tack.util.ViewUtil;

@SuppressLint("CustomSplashScreen")
public class SplashActivity extends MainActivity {

  @Override
  public void onCreate(Bundle bundle) {
    if (Build.VERSION.SDK_INT >= 31) {
      super.onCreate(bundle);

      getSplashScreen().setOnExitAnimationListener(view -> {
        AnimatorSet set = new AnimatorSet();
        set.playTogether(
            ObjectAnimator.ofFloat(view, "alpha", 0),
            ObjectAnimator.ofFloat(view.getIconView(), "alpha", 0)
        );
        set.setDuration(400);
        set.setStartDelay(600);
        set.addListener(new AnimatorListenerAdapter() {
          @Override
          public void onAnimationEnd(@NonNull Animator animation, boolean isReverse) {
            view.remove();
          }
        });
        set.start();
      });
    } else {
      SharedPreferences sharedPrefs = new PrefsUtil(this)
          .checkForMigrations()
          .getSharedPrefs();

      // DARK MODE

      int modeNight = sharedPrefs.getInt(PREF.MODE, DEF.MODE);
      int uiMode = getResources().getConfiguration().uiMode;
      switch (modeNight) {
        case AppCompatDelegate.MODE_NIGHT_NO:
          uiMode = Configuration.UI_MODE_NIGHT_NO;
          break;
        case AppCompatDelegate.MODE_NIGHT_YES:
          uiMode = Configuration.UI_MODE_NIGHT_YES;
          break;
      }
      AppCompatDelegate.setDefaultNightMode(modeNight);
      // Apply config to resources
      Resources resBase = getBaseContext().getResources();
      Configuration configBase = resBase.getConfiguration();
      configBase.uiMode = uiMode;
      resBase.updateConfiguration(configBase, resBase.getDisplayMetrics());

      // THEME

      switch (sharedPrefs.getString(PREF.THEME, DEF.THEME)) {
        case THEME.RED:
          setTheme(R.style.Theme_Tack_Red);
          break;
        case THEME.YELLOW:
          setTheme(R.style.Theme_Tack_Yellow);
          break;
        case THEME.LIME:
          setTheme(R.style.Theme_Tack_Lime);
          break;
        case THEME.GREEN:
          setTheme(R.style.Theme_Tack_Green);
          break;
        case THEME.TURQUOISE:
          setTheme(R.style.Theme_Tack_Turquoise);
          break;
        case THEME.TEAL:
          setTheme(R.style.Theme_Tack_Teal);
          break;
        case THEME.BLUE:
          setTheme(R.style.Theme_Tack_Blue);
          break;
        case THEME.PURPLE:
          setTheme(R.style.Theme_Tack_Purple);
          break;
        default:
          if (DynamicColors.isDynamicColorAvailable()) {
            DynamicColors.applyToActivityIfAvailable(this);
          } else {
            setTheme(R.style.Theme_Tack_Yellow);
          }
          break;
      }

      if (bundle == null) {
        bundle = new Bundle();
      }
      bundle.putBoolean(EXTRA.RUN_AS_SUPER_CLASS, true);
      super.onCreate(bundle);

      new SystemBarBehavior(this).setUp();

      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        LayerDrawable splashContent = (LayerDrawable) ResourcesCompat.getDrawable(
            getResources(), R.drawable.splash_content, getTheme()
        );
        getWindow().getDecorView().setBackground(splashContent);
        try {
          assert splashContent != null;
          ViewUtil.startIcon(splashContent.findDrawableByLayerId(R.id.splash_logo));
          new Handler(Looper.getMainLooper()).postDelayed(this::startNewMainActivity, 900);
        } catch (Exception e) {
          startNewMainActivity();
        }
      } else {
        startNewMainActivity();
      }
    }
  }

  @Override
  protected void attachBaseContext(Context base) {
    if (Build.VERSION.SDK_INT >= 31) {
      super.attachBaseContext(base);
      return;
    }
    SharedPreferences sharedPrefs = new PrefsUtil(base).checkForMigrations().getSharedPrefs();
    // Night mode
    int modeNight = sharedPrefs.getInt(PREF.MODE, DEF.MODE);
    int uiMode = base.getResources().getConfiguration().uiMode;
    switch (modeNight) {
      case AppCompatDelegate.MODE_NIGHT_NO:
        uiMode = Configuration.UI_MODE_NIGHT_NO;
        break;
      case AppCompatDelegate.MODE_NIGHT_YES:
        uiMode = Configuration.UI_MODE_NIGHT_YES;
        break;
    }
    AppCompatDelegate.setDefaultNightMode(modeNight);
    // Apply config to resources
    Resources resources = base.getResources();
    Configuration config = resources.getConfiguration();
    config.uiMode = uiMode;
    resources.updateConfiguration(config, resources.getDisplayMetrics());
    super.attachBaseContext(base.createConfigurationContext(config));
  }

  private void startNewMainActivity() {
    Intent intent = new Intent(this, MainActivity.class);
    intent.addCategory(Intent.CATEGORY_LAUNCHER);
    startActivity(intent);
    overridePendingTransition(0, R.anim.fade_out);
    finish();
  }
}

File: legacy/src/main/kotlin/xyz/zedler/patrick/tack/activity/SplashActivity.kt
/*
 * This file is part of Tack Android.
 *
 * Tack Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Tack Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tack Android. If not, see http://www.gnu.org/licenses/.
 *
 * Copyright (c) 2020-2026 by Patrick Zedler
 */

package xyz.zedler.patrick.tack.activity

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.graphics.drawable.LayerDrawable
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.res.ResourcesCompat
import xyz.zedler.patrick.tack.Constants.DEF
import xyz.zedler.patrick.tack.Constants.EXTRA
import xyz.zedler.patrick.tack.Constants.PREF
import xyz.zedler.patrick.tack.R
import xyz.zedler.patrick.tack.behavior.SystemBarBehavior
import xyz.zedler.patrick.tack.util.PrefsUtil
import xyz.zedler.patrick.tack.util.setTheme
import xyz.zedler.patrick.tack.util.start
import java.time.Instant
import java.time.temporal.ChronoUnit

@SuppressLint("CustomSplashScreen")
class SplashActivity : MainActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    val sharedPrefs = PrefsUtil(this)
      .checkForMigrations()
      .sharedPrefs

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S_V2) {
      super.onCreate(savedInstanceState)

      splashScreen.setOnExitAnimationListener { view ->
        val startTime = view.iconAnimationStart
        val animator = ObjectAnimator.ofFloat(
          view, "alpha", 0f
        )
        animator.duration = 250
        animator.startDelay = startTime?.let {
          (900 - it.until(Instant.now(), ChronoUnit.MILLIS))
            .coerceAtLeast(0)
        } ?: 0
        animator.addListener(object : AnimatorListenerAdapter() {
          override fun onAnimationEnd(animation: Animator, isReverse: Boolean) {
            view.remove()
          }
        })
        animator.start()
      }
    } else {
      // DARK MODE
      val modeNight = sharedPrefs.getInt(PREF.UI_MODE, DEF.UI_MODE)
      var uiMode = resources.configuration.uiMode
      when (modeNight) {
        AppCompatDelegate.MODE_NIGHT_NO -> uiMode = Configuration.UI_MODE_NIGHT_NO
        AppCompatDelegate.MODE_NIGHT_YES -> uiMode = Configuration.UI_MODE_NIGHT_YES
      }
      AppCompatDelegate.setDefaultNightMode(modeNight)
      // Apply config to resources
      val resBase = baseContext.resources
      val configBase = resBase.configuration
      configBase.uiMode = uiMode
      resBase.updateConfiguration(configBase, resBase.displayMetrics)

      // THEME
      setTheme(sharedPrefs)

      val finalBundle = savedInstanceState ?: Bundle()
      finalBundle.putBoolean(EXTRA.RUN_AS_SUPER_CLASS, true)
      super.onCreate(finalBundle)

      SystemBarBehavior(this).setUp()

      val splashContent = ResourcesCompat.getDrawable(
        resources, R.drawable.splash_content, theme
      ) as? LayerDrawable
      window.decorView.background = splashContent
      try {
        checkNotNull(splashContent)
        splashContent.findDrawableByLayerId(R.id.splash_logo).start()
        Handler(Looper.getMainLooper()).postDelayed(
          { startNewMainActivity() }, 600
        )
      } catch (e: Exception) {
        startNewMainActivity()
      }
    }
  }

  override fun attachBaseContext(base: Context) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
      super.attachBaseContext(base)
      return
    }
    val sharedPrefs = PrefsUtil(base).checkForMigrations().sharedPrefs
    // Night mode
    val modeNight = sharedPrefs.getInt(PREF.UI_MODE, DEF.UI_MODE)
    var uiMode = base.resources.configuration.uiMode
    when (modeNight) {
      AppCompatDelegate.MODE_NIGHT_NO -> uiMode = Configuration.UI_MODE_NIGHT_NO
      AppCompatDelegate.MODE_NIGHT_YES -> uiMode = Configuration.UI_MODE_NIGHT_YES
    }
    AppCompatDelegate.setDefaultNightMode(modeNight)
    // Apply config to resources
    val resources = base.resources
    val config = resources.configuration
    config.uiMode = uiMode
    resources.updateConfiguration(config, resources.displayMetrics)
    super.attachBaseContext(base.createConfigurationContext(config))
  }

  private fun startNewMainActivity() {
    val intent = Intent(this, MainActivity::class.java)
    intent.addCategory(Intent.CATEGORY_LAUNCHER)
    startActivity(intent)
    overridePendingTransition(0, R.anim.fade_out)
    finish()
  }
}
